### PR TITLE
STITCH-1074 Add ability for a user to link to any non-anonymous identity

### PR DESCRIPTION
**Pull request by adamchel**
This is pretty much the same as what I did with iOS. Refactored `doLoginRequest()` into `doAuthRequest()`, made the functionality match iOS and JS, and added a test that will be commented out until we have the admin API.

**Drive-bys**
- Added the missing `”provider_id”` field to the `UserProfile` class.
- Changed the `”provider”` field in `UserProfile` to `”provider_type”` to reflect what the server now provides.

### CODE REVIEW

File: stitch/src/main/java/com/mongodb/stitch/android/auth/LinkInfo.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    /**
+     * @return The user this session was created for.
+     */
+    @JsonProperty(value = Fields.USER_ID)
+    public String getUserId() {
+        return _userId;
+    }
+}
```
Comment (jsflax) [inline stitch/src/main/java/com/mongodb/stitch/android/auth/LinkInfo.java]: [nit] Missing new line.

File: stitch/src/main/java/com/mongodb/stitch/android/StitchClient.java [hunk 2]
```diff
@@ -280,52 +283,101 @@ public Void then(@NonNull final Task<String> task) {
         return logout().continueWithTask(new Continuation<Void, Task<String>>() {
             @Override
             public Task<String> then(@NonNull Task<Void> task) throws Exception {
-                return doLoginRequest(authProvider);
+                return doAuthRequest(authProvider,false);
```
Comment (jsflax) [inline stitch/src/main/java/com/mongodb/stitch/android/StitchClient.java]: [nit] Missing space after comma

File: stitch/src/androidTest/java/com/mongodb/stitch/StitchClientServiceTest.kt [hunk 3]
```diff
@@ -270,4 +270,27 @@ class StitchClientServiceTest {
         assertFalse(mlsStitchClient.isAuthenticated)
         assertEquals(mlsStitchClient.loggedInProviderType, "")
     }
+
+    // NOTE: This test works locally when logging in with an email identity not associated with a Stitch user, but with our
```
Comment (jsflax) [inline stitch/src/androidTest/java/com/mongodb/stitch/StitchClientServiceTest.kt]: [nit] We should add a TODO here, so IntelliJ catches it.

Comment (edaniels) [reply stitch/src/androidTest/java/com/mongodb/stitch/StitchClientServiceTest.kt]: Can we also add tickets to go back to this?

File: stitch/src/main/java/com/mongodb/stitch/android/StitchClient.java [hunk 4]
```diff
@@ -280,52 +283,101 @@ public Void then(@NonNull final Task<String> task) {
...
+                    new StitchException.StitchClientException(
+                            "Must be authenticated to link a user to new identity."
+                    )
+            );
+        }
+
+        return doAuthRequest(authProvider, true);
+    }
+
+    private Task<String> doAuthRequest(final AuthProvider authProvider, final boolean linking) {
```
Comment (edaniels) [inline stitch/src/main/java/com/mongodb/stitch/android/StitchClient.java]: [nit] maybe `shouldLink`?

**Pull request merged**